### PR TITLE
Filter packages listed in notebooks

### PR DESCRIPTION
**Pull request by frankharkins**
At the top of each notebook, we list the packages needed to run every notebook in this repo. This PR analyses the notebook filter out any packages not required by that notebook.

I've also re-run three notebooks to demonstrate, the rest will be updated in the fortnightly cron job.

### CODE REVIEW

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py [hunk 1]
```diff
@@ -37,15 +43,43 @@ def post_process_notebook(nb: nbformat.NotebookNode) -> nbformat.NotebookNode:
...
+      if module not in sys.stdlib_module_names
+    ])
     package_versions = "\n".join(
-        line for line in requirements_file.split("\n") if not line.startswith("#")
+        line for line in requirements_txt.split("\n")
+        if re.split('[\\[~=]', line)[0].strip() in packages
     )
     return package_versions.strip()
+
+def get_used_modules(source: str) -> Iterable[str]:
```
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:70]: You could instead take the arg as a `list[str]`. Note that you joined it with `\n` up above, then split, then rejoin again.

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py [hunk 2]
```diff
@@ -37,15 +43,43 @@ def post_process_notebook(nb: nbformat.NotebookNode) -> nbformat.NotebookNode:
...
+    # Remove Jupyter magics
+    source = "\n".join([
```
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:72]: It's best practice to not shadow the argument with the variable name.

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py [hunk 3]
```diff
@@ -37,15 +43,43 @@ def post_process_notebook(nb: nbformat.NotebookNode) -> nbformat.NotebookNode:
...
+        line for line in source.split("\n")
+        if not line.strip().startswith("%")
+    ])
+    for node in ast.iter_child_nodes(ast.parse(source=source, filename="_.py")):
+        if isinstance(node, ast.Import):
+            for module in node.names:
+                yield module.name.split(".")[0]
```
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:79]: FYI there are certain packages that expose a module like `overall_namespace.my_pkg_namespace`, and another dep exposes `overall_namespace.another_pkg_namespace`. We don't have any deps that do that though, so I am fine with this simpler approach.

I doubt we will encounter this problem in the future; it's not a common pattern. So I think this is fine to not have a comment.

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py [hunk 4]
```diff
@@ -37,15 +43,43 @@ def post_process_notebook(nb: nbformat.NotebookNode) -> nbformat.NotebookNode:
...
     )
     return package_versions.strip()
+
+def get_used_modules(source: str) -> Iterable[str]:
+    # Remove Jupyter magics
+    source = "\n".join([
+        line for line in source.split("\n")
+        if not line.strip().startswith("%")
+    ])
+    for node in ast.iter_child_nodes(ast.parse(source=source, filename="_.py")):
```
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:76]: The AST parsing can fail if there are syntax errors:

```
>>> ast.parse("imprtttttt sys", filename="_.py")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/arellano/.pyenv/versions/3.9.17/lib/python3.9/ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
  File "_.py", line 1
    imprtttttt sys
               ^
SyntaxError: invalid syntax
```

Generally, I'd say that's fine, but I think it would break nb-tester for every other notebook that is valid because it would crash the overall nb-tester program. So, we should probably catch it. However, we do want to fail the overall run, only not crash it. So maybe return a `Result` that preserves the exception?

I do think this is probably worth fixing because the error would be cryptic otherwise.

It'd be good to test this edge case.

--

Another small tip is to make the call `ast.parse(source=source, filename="_.py")` a standalone line and save as a variable like `ast_tree`. That will make the stacktrace more clear that that specific function call was the problem.

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py [hunk 5]
```diff
@@ -37,15 +43,43 @@ def post_process_notebook(nb: nbformat.NotebookNode) -> nbformat.NotebookNode:
...
+        if isinstance(node, ast.Import):
+            for module in node.names:
+                yield module.name.split(".")[0]
+        elif isinstance(node, ast.ImportFrom):
+            if node.module is not None:
+                yield node.module.split(".")[0]
+
+def flatten_to_list(i: Iterable[Iterable]) -> list:
```
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:84]: Imo it's fine to call this `flatten` because the type hint already makes clear it's a list and it's a sensible default. I wish Python std lib implemented it.

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py [hunk 6]
```diff
@@ -10,7 +10,13 @@
 # notice, and modified files need to carry a notice indicating that they have
 # been altered from the originals.
 
+import ast
+import re
+import sys
+import importlib
+import itertools
 from pathlib import Path
+from typing_extensions import Iterable
```
Comment (Eric-Arellano) [inline scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:19]: This should be `collections.abc`.

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py (hunk 2 above)
Comment (Eric-Arellano) [reply scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:72]: Also FYI no need to use the list comprehension with the outer `[]`. You can use a generator expression by removing the `[]`. That is lazy (uses an iterator), rather than storing the whole memory in list.

**Review by Eric-Arellano — COMMENTED** [review]
Awesome job, Frank! Good job taking initiative and very readable code :) 

This is a fun PR for me because you re-created dependency inference, a key feature of Pantsbuild I worked on.

--

Technically, it would be more robust to have a distinct venv per notebook execution to be confident that we didn't miss any dependencies. With the current approach, it is theoretically possible for a dep to be necessary but not inferred.

However, it is not necessary to make that change. I'm confident in your dependency inference code and it's not worth the complexity to have multiple venvs (Pantsbuild would do it for us, but that's out of scope).

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py (hunk 3 above)
Comment (frankharkins) [reply scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:79]: I agree it's unlikely, however, if that does happen, we'll list both packages in the requirements as we include all packages from [`importlib.metadata.package_distributions()`](https://docs.python.org/3/library/importlib.metadata.html#package-distributions). I left a comment about it in `get_package_versions`.

**Comment by frankharkins** [comment]
Thanks!

> Technically, it would be more robust to have a distinct venv per notebook execution...

This was my first thought, however, it's gonna be messy making that work (listing problems for posterity):

* AFAIK there's no Python API to create a venv, so we'd have to call out to subprocesses to create the venvs and link them to IPython kernels. We'd also need to delete each kernel after running.
* nbconvert isn't super robust and we've already hit weird flakes when running lots of notebooks (e.g. #2387). Running ~60 different kernels is likely to introduce more strange bugs.
* We'd have to list the individual requirements for each notebook and find a way to update them all at the same time (e.g. with dependabot).

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py (hunk 4 above)
Comment (frankharkins) [reply scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:77]: We can assume the code is valid as we only post-process notebooks after successful execution. However, I've added `SyntaxError` to the list of expected exceptions so we don't crash the whole script if something unexpected happens (see d7b618bb0262fb9563851baf48f8a1deead18b70).

File: scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py (hunk 1 above)
Comment (frankharkins) [reply scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py:71]: The first time we join (L47), we're joining cells, which can contain many lines. We could split those cells and flatten the list, but I think joining there is better because I'd expect "source code" to be a single string rather than a list.
